general/globals.hpp

```diff
@@ -114,4 +114,16 @@ void SetGlobalMPI_Comm(MPI_Comm comm);
 #define MFEM_THREAD_LOCAL thread_local
 
 
+// MFEM_DEPRECATED macro to mark obsolete functions and methods
+// see https://stackoverflow.com/questions/295120/c-mark-as-deprecated
+#if defined(__GNUC__) || defined(__clang__)
+#define MFEM_DEPRECATED __attribute__((deprecated))
+#elif defined(_MSC_VER)
+#define MFEM_DEPRECATED __declspec(deprecated)
+#else
+#pragma message("WARNING: You need to implement MFEM_DEPRECATED for this compiler")
+#define MFEM_DEPRECATED
+#endif
+
```

Comment:
I am about to merge this, but I do have some concerns about this new macro:

There is already an existing way to handle deprecated methods in the library: by adding `(DEPRECATED)` in the beginning of the documentation and/or the Doxygen `@deprecated` command, see https://github.com/mfem/mfem/search?q=DEPRECATED&unscoped_q=DEPRECATED. 

Isn't this alternative way confusing things?

From user perspective, marking the method as deprecated in the documentation seems preferable, but maybe the compiler attribute has distinctive advantages? If so, what happens e.g. with `xlC`?

I don't have strong feelings about this one way or the other, but I'd like to see a unified approach.

Reply:
I didn't realize we already had a convention to mark deprecated methods. The `__attribute__` (or similar) approach has the advantage that the user gets a compiler warning when such function is used. I'm fine with using the "softer" deprecation by documentation, if you prefer. 

For `xlC` we would have to figure out if this feature is supported and define the macro accordingly, or disable it.

Reply:
My two cents (andI don't have strong feelings about this): I don't think the softer approach is very effective because it's so easy for users to ignore.  I'm in favor of @jakubcerveny 's use of compiler warnings when available.

Reply:
Okay, so how do we unify?

1. Switch all deprecated instances to the MFEM_DEPRECATED only (and remove current comments)
2. Switch to `DEPRECATED` and `@deprecated` comments only (and remove `MFEM_DEPRECATED`)
3. Require both `MFEM_DEPRECATED` and the `DEPRECATED`/`@deprecated` comments

Based on the above comments it looks like @jakubcerveny and @mlstowell prefer 1 or 3.

I vote for 3.

Reply:
I vote for 3, `MFEM_DEPRECATED` and one of `DEPRECATED`/`@deprecated` comments.

Reply:
I vote for 3.

Reply:
@jakubcerveny, can you then make the change 3 in this PR?

Reply:
Done, added the macro wherever I found `@deprecated` in the comments. 

One exception was `Device::GetMemoryClass` and `Device::GetMemoryType`, which still seem widely used and I didn't want to start changing a lot of code.